File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxi_dia=0;
    int solve(TreeNode* root){
        if(!root)return 0;
        int left=solve(root->left);
        int right=solve(root->right);
        maxi_dia=max(maxi_dia,left+right);
        return max(left,right)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int i=solve(root);
        return maxi_dia;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxi=-1;
    int dfs(TreeNode root){
        if(root==null)return 0;
        int left=0;
        int right=0;
        if(root.left!=null){
            left=dfs(root.left);
        }
        if(root.right!=null){
            right=dfs(root.right);
        }
        int dia=left+right;
        if(dia>maxi){
            maxi=dia;
        }
        if(left>right){
            return 1+left;
        }else return 1+right;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        int dummy=dfs(root);
        return maxi;
    }
}